### ISSUE TITLE
Specify DSSE Signature Algorithm / Format in the Protocol or as a Parameter

### ISSUE THREAD

**Issue by CrossedSecurity** [secure-systems-lab/dsse#49]
In the current state, we have DSSE's that contain a signature and the information needed to generate the PAE(message) that gets signed. When a DSSE Verifier is created, you must specify a signature algorithm (e.g. ECDSA), a few other parameters, and a signature encoding scheme (e.g. DER, IEEE_P1363). Unfortunately, if you attempt a DSSE verification using the incorrect signature encoding, it is unlikely that the crypto library is going to tell you that, and it's rather painful to debug.

To avoid ambiguities around what algorithm/parameters/signature encoding was used to sign a DSSE's PAE(message), we should either add these as a requirement in the spec, or provide a place to specify which of them were used. If we decide to add this to the protocol, which should consider which systems are able to produce the encoding format, and/or how difficult of a task converting between them is.

**Comment by trishankatdatadog** [comment]
Hmm, pretty sure we decided to delegate this job to the _application_ using DSSE. For example, both TUF and in-toto (for which DSSE was proposed) define signature algorithm parameters within the _signed_ message.

@MarkLodato @SantiagoTorres thoughts here?

**Comment by MarkLodato** [comment]
Oops, as I re-read @CrossedSecurity's initial post, I see now that he did ask about algorithm. I was basing my comment off our earlier private conversation. I think we should scope this issue to just encoding, since as mentioned above, we had previously decided not to transmit algorithm in-band.

**Comment by trishankatdatadog** [comment]
> To clarify, this is purely about encoding, not the algorithm.

I see, thanks for the clarification!

> * (A) Communicate the encoding as part of the public key parameters.

My vote is for this: consistent with our approach for the rest of the metadata (e.g., algorithm parameters) and also allows for flexibility.

**Comment by CrossedSecurity** [comment]
Right, sorry. I was thinking that if we were to specify or recommend the signature encoding somewhere, the we'd have to pair that with whichever signing algo was used. (RSA doesn't use DER signatures, for instance). But yea we could just infer that from the public key.

(A) is a good solution but could present an issue if somebody signs something with that key using the incorrect signature encoding. This is error-prone because there is no hard-link between what signature encoding is published and what actually happens during signing. At least if this information is published, it's a little easier to debug when that happens.

If we're already doing this with the other algorithm parameters, this would be the easiest path forward.

(C) would also be fine imo, with the caveat being that some KMS's only output certain encodings (e.g. GCP CloudKMS only outputs DER).

**Comment by trishankatdatadog** [comment]
> (A) is a good solution but could present an issue if somebody signs something with that key using the incorrect signature encoding. This is error-prone because there is no hard-link between what signature encoding is published and what actually happens during signing. At least if this information is published, it's a little easier to debug when that happens.
> 
> If we're already doing this with the other algorithm parameters, this would be the easiest path forward.

Precisely: if the encoding is published and signed, then that's what the signer should follow. WDYT?
